subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,630 @@
+
+	if (ack) { /* reply on the action set command */
+		send_scheduler_action_status(model, ctx, idx);
+	}
+}
+
+static void handle_scheduler_get(struct bt_mesh_model *mod,
+				 struct bt_mesh_msg_ctx *ctx,
+				 struct net_buf_simple *buf)
+{
+	send_scheduler_status(mod, ctx);
+}
+
+static void handle_scheduler_action_get(struct bt_mesh_model *mod,
+					struct bt_mesh_msg_ctx *ctx,
+					struct net_buf_simple *buf)
+{
+	uint8_t idx = net_buf_simple_pull_u8(buf);
+
+	if (idx > 0xf) {
+		return;
+	}
+
+	send_scheduler_action_status(mod, ctx, idx);
+}
+
+static void handle_scheduler_action_set(struct bt_mesh_model *mod,
+					struct bt_mesh_msg_ctx *ctx,
+					struct net_buf_simple *buf)
+{
+	action_set(mod, ctx, buf, true);
+}
+
+static void handle_scheduler_action_set_unack(struct bt_mesh_model *mod,
+					      struct bt_mesh_msg_ctx *ctx,
+					      struct net_buf_simple *buf)
+{
+	action_set(mod, ctx, buf, false);
+}
+
+const struct bt_mesh_model_op _bt_mesh_scheduler_srv_op[] = {
+	{ BT_MESH_SCHEDULER_OP_GET,
+	  BT_MESH_SCHEDULER_MSG_LEN_GET,
+	  handle_scheduler_get },
+	{ BT_MESH_SCHEDULER_OP_ACTION_GET,
+	  BT_MESH_SCHEDULER_MSG_LEN_ACTION_GET,
+	  handle_scheduler_action_get },
+	  BT_MESH_MODEL_OP_END,
+};
+
+const struct bt_mesh_model_op _bt_mesh_scheduler_setup_srv_op[] = {
+	{ BT_MESH_SCHEDULER_OP_ACTION_SET,
+	  BT_MESH_SCHEDULER_MSG_LEN_ACTION_SET,
+	  handle_scheduler_action_set },
+	{ BT_MESH_SCHEDULER_OP_ACTION_SET_UNACK,
+	  BT_MESH_SCHEDULER_MSG_LEN_ACTION_SET,
+	  handle_scheduler_action_set_unack },
+	  BT_MESH_MODEL_OP_END,
+};
+
+static int scheduler_srv_init(struct bt_mesh_model *mod)
+{
+	struct bt_mesh_scheduler_srv *srv = mod->user_data;
+
+	srv->mod = mod;
+	net_buf_simple_init(mod->pub->msg, 0);
+	srv->data.status_bitmap = 0;
+	memset(&srv->data.sch_reg, 0, sizeof(srv->data.sch_reg));
+
+	struct bt_mesh_model *time_mod =
+			bt_mesh_model_find(bt_mesh_model_elem(mod),
+			BT_MESH_MODEL_ID_TIME_SRV);
+	if (time_mod == NULL) {
+		return -ECANCELED;
+	}
+	srv->time_srv = (struct bt_mesh_time_srv *)time_mod->user_data;
+
+	if (IS_ENABLED(CONFIG_BT_MESH_MODEL_EXTENSIONS)) {
+		/* Model extensions:
+		 * To simplify the model extension tree, we're flipping the
+		 * relationship between the scheduler server and the scheduler
+		 * setup server. In the specification, the scheduler setup
+		 * server extends the scheduler server, which is the opposite
+		 * of what we're doing here. This makes no difference for
+		 * the mesh stack, but it makes it a lot easier to extend
+		 * this model, as we won't have to support multiple extenders.
+		 */
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_SCHEDULER_SETUP_SRV));
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_SCENE_SRV));
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_GEN_POWER_ONOFF_SRV));
```

Comment:
5.3.6.1 Description
The Scheduler Setup Server model is used to support the configuration functionality of a node that
can store a schedule.
The Scheduler Setup Server is a main model that extends and corresponds with the Scheduler Server
model (see Section 5.3.5), and extends the Scene Setup Server (see Section 5.3.4) and the Generic
Power OnOff Setup Server (see Section 3.3.5) models.

Reply:
You're right. I posted a discussion in our team chat, I think this introduces a couple of issues.
The problem mentioned in the extension comment that's copied in all the setup server models also needs to be considered; we don't support models being extended by multiple other models at the moment. This can be addressed in upstream by switching to a non-hierarchical extension tree, like a circular linked list or something. We'll need to find a way to preserve the hierarchy later though.